### PR TITLE
API/YAML: Support using env vars to define file_mounts.

### LINKED ISSUE #2093: Use environment variable to define file mount
Hi and thanks for this great project !

I am in a situation where I would like to mount some directory in a bucket which changes based on the value of an environment variable. I am unsure how to do this with SkyPilot as defining a file mount would typically require that I specify the corresponding URL in my task YAML file beforehand.

Put simply I'd like to do
```yaml
file_mounts:
  /some/folder:
    source: $URL_IN_MY_ENVVAR
    mode: MOUNT
```

And be able to set `URL_IN_MY_ENVVAR ` at runtime.
Am I missing something or would it be a feature that may be of interest?

### PR DESCRIPTION
**Pull request by concretevitamin**
```
# Use env vars in the "name" (bucket name) and "source" (local dir/file to
# upload) fields.
#
# Both syntaxes work: ${MY_BUCKET} and $MY_BUCKET.
file_mounts:
  /mydir:
    name: ${MY_BUCKET}  # Name of the bucket.
    store: gcs
    mode: MOUNT

  /another-dir:
    name: ${MY_BUCKET}-2
    source: ["~/${MY_LOCAL_PATH}"]
    store: gcs
    mode: MOUNT
```
Overriding with `sky launch --env` works too.

Also fixes `sky launch` parsing errors: 
- Do not load YAML twice (which creates storage twice)
- Properly load empty / name-only YAMLs

Fixes #2093.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - `pytest tests/test_smoke.py::test_using_file_mounts_with_env_vars`
  - `pytest tests/test_optimizer_dryruns.py`
- [x] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/task.py [hunk 1]
```diff
@@ -68,6 +68,45 @@ def _is_valid_env_var(name: str) -> bool:
...
+        - ${ENV}
+        - $ENV
+    where <ENV> must appear in task.envs.
+
+    Example:
+        {'mode': 'MOUNT', 'name': '${GSBUCKET}', 'store': 'gcs'}
+        -> {mode: 'MOUNT', name: 'mybucket', store: 'gcs'}
+    """
+
+    # TODO(zongheng): support ${ENV:-default}?
```
Comment (Michaelvll) [inline sky/task.py:89]: I think it is ok to not support this, as the user can specify the default value for the env in the yaml file. We can keep this though.

File: sky/task.py [hunk 2]
```diff
@@ -68,6 +68,45 @@ def _is_valid_env_var(name: str) -> bool:
...
+    def _substitute_env_vars(target: Union[str, List[str]],
+                             envs: Dict[str, str]) -> Union[str, List[str]]:
+        for key, value in envs.items():
+            pattern = r'\$\{?\b' + key + r'\b\}?'
```
Comment (Michaelvll) [inline sky/task.py:93]: We may need to add a check for the key of the `task.envs` somewhere in our code, i.e. we need to make sure the `key ~ '[a-zA-Z][a-zA-Z0-9]*'`

File: sky/task.py [hunk 3]
```diff
@@ -269,10 +320,12 @@ def from_yaml_config(config: Dict[str, Any]) -> 'Task':
...
             mount_path = storage[0]
-            assert mount_path, \
-                'Storage mount path cannot be empty.'
+            assert mount_path, 'Storage mount path cannot be empty.'
             try:
-                storage_obj = storage_lib.Storage.from_yaml_config(storage[1])
+                storage_config = _fill_in_env_vars_in_storage_config(
+                    storage[1], task.envs)
+                storage_obj = storage_lib.Storage.from_yaml_config(
+                    storage_config)
```
Comment (Michaelvll) [inline sky/task.py:328]: Instead of overriding the `storage_config`, should we just replace all the env var in the `file_mounts` section? For example, in L285 we do:
```
if config.get('file_mounts') is not None:
    file_mounts_str = yaml.dumps(config['file_mounts'])
    new_file_mounts_str = _replace_env_var(file_mounts_str, config.get('envs', {}))
    config['file_mounts'] = yaml.loads(new_file_mounts_str)
```

With that, the user can do something like the following:
```
file_mounts:
  /model_path/llama-${SIZE}b: s3://llama-weights/llama-${SIZE}b
```

File: tests/test_smoke.py [hunk 4]
```diff
@@ -738,6 +738,28 @@ def test_scp_file_mounts():
     run_one_test(test)
 
 
+def test_using_file_mounts_with_env_vars(generic_cloud: str):
+    name = _get_cluster_name()
+    test_commands = [
+        *storage_setup_commands,
+        (f'sky launch -y -c {name} --cpus 2+ --cloud {generic_cloud} '
+         'examples/using_file_mounts_with_env_vars.yaml'),
```
Comment (Michaelvll) [inline tests/test_smoke.py:746]: This file seems not included?

File: sky/cli.py [hunk 5]
```diff
@@ -1041,7 +1045,7 @@ def _make_task_or_dag_from_entrypoint_with_overrides(
     if is_yaml:
         assert entrypoint is not None
         usage_lib.messages.usage.update_user_task_yaml(entrypoint)
-        dag = dag_utils.load_chain_dag_from_yaml(entrypoint)
+        dag = dag_utils.load_chain_dag_from_yaml(entrypoint, env_overrides=env)
```
Comment (Michaelvll) [inline sky/cli.py:1048]: Our original behavior is to ignore all the overrides, including `env`. Otherwise, the semantics will be a bit weird, i.e. overriding envs in all the tasks or just overriding the first task. Maybe we don't need this `env_overrides` option should not take effect for a dag with more than 1 task?

**Review by Michaelvll — COMMENTED** [review]
Thanks for adding this @concretevitamin! This should make our YAML interface more flexible. Left several questions.

File: sky/cli.py (hunk 5 above)
Comment (concretevitamin) [reply sky/cli.py:1052]: Sounds good. Done.

File: sky/task.py (hunk 1 above)
Comment (concretevitamin) [reply sky/task.py:89]: Sg. From user issue it seems that this syntax is applied in the run section. Keeping the TODO.

File: tests/test_smoke.py (hunk 4 above)
Comment (concretevitamin) [reply tests/test_smoke.py:746]: Oops, added.

File: sky/task.py (hunk 2 above)
Comment (concretevitamin) [reply sky/task.py:93]: This will get checked by storage.py's validate_name().

File: sky/task.py (hunk 3 above)
Comment (concretevitamin) [reply sky/task.py:328]: One worry with string replacement is that it'll make some weird things "work":
```
file_mounts:
  /path:
    mo${empty_envvar}de: xxxx
```
I think we should make replacement more structured.

That said, supporting env vars in src/dst seems a useful case. How about we support it when users ask?

Comment (Michaelvll) [reply sky/task.py:328]: Sounds good, though I still think the replacement for `dst` and `src` might be quite important, as that was one thing we would like to have for the Vicuna for downloading the correct llama weight from the bucket based on the envvar.

For the problem you mentioned, we can do the schema validation before the replacement happens to avoid that issue.

File: sky/task.py (hunk 2 above)
Comment (Michaelvll) [reply sky/task.py:93]: Ahh, I was trying to say that if the user provides a envvar like the following:
```
envs:
  hello world: some_value

run: |
  echo ${hello world}
```

The behavior will be different, if this env var is used in the `file_mounts` section.
```
file_mounts:
  /dst:
    source: ${hello world}
```

File: sky/task.py (hunk 3 above)
Comment (concretevitamin) [reply sky/task.py]: Good call! Updated as such.

File: sky/task.py (hunk 2 above)
Comment (concretevitamin) [reply sky/task.py]: Good call! Done.

File: sky/task.py [hunk 6]
```diff
@@ -68,6 +69,35 @@ def _is_valid_env_var(name: str) -> bool:
...
+
+    Env vars of the following forms are detected:
+        - ${ENV}
+        - $ENV
+    where <ENV> must appear in task.envs.
+    """
+    # TODO(zongheng): support ${ENV:-default}?
+    file_mounts_str = json.dumps(file_mounts)
+    for key, value in task_envs.items():
+        pattern = r'\$\{?\b' + key + r'\b\}?'
```
Comment (Michaelvll) [inline sky/task.py:95]: nit: it seems bash replaces the envvar by looking for the envvar in the string first and match it with the existing envvar, while here we do it in a reversed way. That will cause a slight inconsistency in the behavior:
```
HELLO_WORLD=1
echo $HELLO_WORLD_1
```
The block above will output an empty string, but in our implementation, we will output `1_1`.

Probably, an alternative way to do this is:
```
# Create a replacement function
def replace_var(match):
    var_name = match.group(1)  # Now directly accessing the variable name
    return task_envs.get(var_name, match.group(0))  # If the variable isn't in the dictionary, return it unchanged

pattern = r'\$\{?\b([a-zA-Z_][a-zA-Z0-9_]*)\b\}?'
# Use re.sub with the replacement function
result = re.sub(pattern, replace_var, file_mounts_str)
```

This is a very corner case, we can also leave it as is to see how people feel about it.

File: tests/test_smoke.py [hunk 7]
```diff
@@ -738,6 +738,28 @@ def test_scp_file_mounts():
     run_one_test(test)
 
 
+def test_using_file_mounts_with_env_vars(generic_cloud: str):
+    name = _get_cluster_name()
+    test_commands = [
+        *storage_setup_commands,
+        (f'sky launch -y -c {name} --cpus 2+ --cloud {generic_cloud} '
```
Comment (Michaelvll) [inline tests/test_smoke.py:745]: nice, we should have all our tests to be `--cpus 2+` in the future to save cost. ; )

**Review by Michaelvll — APPROVED** [review]
Thanks for the fix @concretevitamin! The code looks good to me with a small corner case that we may need to be careful.

File: sky/task.py (hunk 6 above)
Comment (concretevitamin) [reply sky/task.py]: Great catch! Used this & reran the smoke test.

**Comment by concretevitamin** [comment]
- [x]  All smoke tests: `pytest tests/test_smoke.py`
